core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java

```diff
@@ -11,20 +11,144 @@
 public class AsyncCacheRequestManager {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncCacheRequestManager.class);
+
+  private final StorageTierAssoc mStorageTierAssoc = new WorkerStorageTierAssoc();
+  /** Executor service for execute the async cache tasks. */
+  private final ExecutorService mAsyncCacheExecutor;
+  /** The block worker. */
+  private final BlockWorker mBlockWorker;
+  private final Object mLock = new Object();
+  @GuardedBy("mLock")
+  private final Set<Long> mPendingRequests;
+  private final WorkerNetAddress mLocalWorkerAddress;
+
+  public AsyncCacheRequestManager(ExecutorService service, BlockWorker blockWorker) {
+    mAsyncCacheExecutor = service;
+    mBlockWorker = blockWorker;
+    mPendingRequests = new HashSet<>();
+    try {
+      mLocalWorkerAddress = FileSystemContext.INSTANCE.getLocalWorker();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
   /**
    * Handles a request to cache a block asynchronously. This is a non-blocking call.
    *
    * @param request the async cache request
    * fields will be available
    */
   public void submitRequest(Protocol.AsyncCacheRequest request) {
+    long blockId = request.getBlockId();
+    synchronized (mLock) {
+      if (mPendingRequests.contains(blockId)) {
+        // This block is already planned.
+        return;
+      }
+      mPendingRequests.add(request.getBlockId());
+    }
+    mAsyncCacheExecutor.submit(() -> {
+      Protocol.OpenUfsBlockOptions openUfsBlockOptions = request.getOpenUfsBlockOptions();
+      long sessionId = IdUtils.createSessionId();
+      long blockSize = openUfsBlockOptions.getBlockSize();
+      boolean isSourceLocal = mLocalWorkerAddress.getHost().equals(request.getSourceHost())
+          && mLocalWorkerAddress.getRpcPort() == request.getSourcePort();
+      // Depends on the request, cache the target block from different sources
+      try {
+        if (isSourceLocal) {
+          cacheBlockFromUfs(sessionId, blockId, blockSize, openUfsBlockOptions);
+        } else {
+          InetSocketAddress sourceAddress =
+              new InetSocketAddress(request.getSourceHost(), request.getSourcePort());
+          cacheBlockFromRemoteWorker(sessionId, blockId, blockSize, sourceAddress,
+              openUfsBlockOptions);
+        }
+      } catch (BlockAlreadyExistsException e) {
+        // It is already cached
+      } catch (AlluxioException | IOException e) {
+        // This is only best effort
+        LOG.warn("Failed to async cache block {}: {}", blockId, e.getMessage());
+      }
+      mPendingRequests.remove(request.getBlockId());
+    });
+  }
 
+  /**
+   * Caches the block via the local worker to read from UFS.
+   *
+   * @param sessionId session ID
+   * @param blockId block ID
+   * @param blockSize block size
+   * @param openUfsBlockOptions options to open the UFS file
+   */
+  private void cacheBlockFromUfs(long sessionId, long blockId, long blockSize,
```

Comment:
Do we want to return boolean to indicate success/failure?

Reply:
Also, I think we need to call something like `commitUfsBlock`, since closing the reader does not handle the block commit.

Reply:
-  `closeUfsBlock` is added
- exceptions are thrown to indicate failures, and handled together in `submitRequest` to simplify the code. Javadoc is updated to reflect the failure cases.